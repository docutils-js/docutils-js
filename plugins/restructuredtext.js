require("@babel/polyfill");

const env = require('jsdoc/env');

const docutils = require('../lib/index');
const parse = docutils.parse;
const pojoTranslate = docutils.pojoTranslate;
const htmlTranslate = docutils.htmlTranslate;
const StackTrace = require('stacktrace-js');

const config = (env && env.conf && env.conf.restructuredtext) || {};
const defaultTags = [
    'author',
    'classdesc',
    'description',
    'exceptions',
    'params',
    'properties',
    'returns',
    'see',
    'summary',
];

const hasOwnProp = Object.prototype.hasOwnProperty;
let tags = [];
let excludeTags = [];
let handleFile = false;

function shouldProcessString(tagName, text) {
    const shouldProcess = true;
    return shouldProcess;
}

function process(doclet) {
    tags.forEach((tag) => {
        if (!hasOwnProp.call(doclet, tag)) {
            return;
        }
        if (typeof doclet[tag] === 'string' && shouldProcessString(tag, doclet[tag])) {
            console.log(`${tag}:\n---\n${doclet[tag]}\n---`);
            try {
                const document = parse(doclet[tag]);
                const r = htmlTranslate(document);
                console.log(r);
                if(typeof r === 'undefined') {
                    throw Error('bad result');
                }
                doclet[tag] = r;//'<p>' + doclet.longname + '</p><div style="border: 1px solid black">' + r + '</div>';
            } catch(e) {
                throw e;
//                StackTrace.fromError(e).then(x => doclet[tag] = x.map((sf) => sf.toString()).join('<br>\n')).catch(e => undefined);
                doclet[tag] = StackTrace.getSync().map((sf) => sf.toString()).join('<br>\n');

            }
        } else if ( Array.isArray(doclet[tag]) ) {
            doclet[tag].forEach((value, index, original) => {
                const inner = {};

                inner[tag] = value;
                process(inner);
                original[index] = inner[tag];
            });
        } else if (doclet[tag]) {
            process(doclet[tag]);
        }
    });
}


// set up the list of "tags" (properties) to process
if (config.tags) {
    tags = config.tags.slice();
}
// set up the list of default tags to exclude from processing
if (config.excludeTags) {
    excludeTags = config.excludeTags.slice();
}
defaultTags.forEach(tag => {
    if (!excludeTags.includes(tag) && !tags.includes(tag)) {
        tags.push(tag);
    }
});


exports.defineTags = (dictionary) => {
    dictionary.defineTag('rst', { onTagged: (doclet, tag) => {
        doclet.isRest = true;
        //console.log(doclet);
    }});
};

exports.handlers = {
    newDoclet: ({ doclet }) => {
        console.log(doclet);
        if(typeof doclet.unknisRest !== 'undefined') {
            console.log(`isrest: ${doclet.isRest}`);
        }
        if(handleFile || doclet.isRest) {
            process(doclet);
        }
    },
    fileBegin: (e) => {
        handleFile = false;
    },
    symbolFound: (e) => {
        if(e.code.name === '__docformat__' && typeof e.code.value === 'string' && e.code.value.toLowerCase() == 'restructuredtext') {
            handleFile = true;
        }
    },
};
