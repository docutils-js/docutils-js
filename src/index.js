/* eslint-disable-next-line import/prefer-default-export */
import parse from './parser';
import { StringOutput } from './io';

export const __version__ = '0.14js';

export { parse, StringOutput };
