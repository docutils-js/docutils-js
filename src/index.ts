/* eslint-disable-next-line import/prefer-default-export */
export { parse } from './parse';
export { StringOutput, StringInput } from './io';
export { StandaloneReader } from './readers/standalone';
export { newDocument } from './newDocument';
import * as nodes from './nodes';
import Writer from './Writer';
import Transform from './Transform';
export { Reader } from './Reader';
export { Publisher } from './Publisher'
export { defaults } from '../gen/defaults';
export { pojoTranslate } from './fn/pojoTranslate';
export { htmlTranslate } from './fn/htmlTranslate';
export { RSTParser } from './parsers/restructuredtext';
export { XMLWriter } from './writers/xml'

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

export const __version__ = '0.14js';

export {
    nodes, Writer, Transform,
};
