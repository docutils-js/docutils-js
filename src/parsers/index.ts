import Parser from '../Parser';
import RestructuredTextParser from './restructuredtext';
import { ParserConsructor } from "../types";

function getParserClass(parserName: string): ParserConsructor {
    if (parserName === 'restructuredtext') {
        return RestructuredTextParser;
    }
    throw new Error('');
//    return require(`./${parserName}.js`).default;
}

export default {
    getParserClass,
    Parser,
};
