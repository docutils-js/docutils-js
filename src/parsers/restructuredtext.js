import BaseParser from '../Parser';

export class Parser extends BaseParser {
    parse() {
    }
    
}

export default {
    Parser,
}

