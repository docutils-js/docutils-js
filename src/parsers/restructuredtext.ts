import BaseParser from '../Parser';
import * as statemachine from '../StateMachine';
import RSTStateMachine from './rst/RSTStateMachine';
import StateFactory from './rst/StateFactory';
import {Document, ParserArgs} from "../types";
import { InlinerInterface } from "./rst/types";
import { InvalidStateError } from "../Exceptions";

class Parser extends BaseParser {
    private inliner?: InlinerInterface;
    private initialState: string;
    private stateMachine?: RSTStateMachine;
    public constructor(args: ParserArgs = {}) {
        super(args);
        this.configSection = 'restructuredtext parser';
        this.configSectionDependencies = ['parsers'];
        if (args.rfc2822) {
            this.initialState = 'RFC2822Body';
        } else {
            this.initialState = 'Body';
        }

        if(args.inliner !== undefined) {
            this.inliner = args.inliner;
        }
    }

    public parse(inputstring: string, document: Document): void {
        if(!inputstring.split) {
            throw new Error('not a string');
        }
        if (!inputstring) {
            throw new Error('need input for rst parser');
        }

        this.setupParse(inputstring, document);

        this.stateMachine = new RSTStateMachine({
            stateFactory: new StateFactory(),
            initialState: this.initialState,
            debugFn: this.debugFn,
            debug: document.reporter.debugFlag,
        });
        if(document.settings.docutilsParsersRstParser === undefined) {
            throw new InvalidStateError(('need document and config for rstparser'));
        }
        let tabWidth = document.settings.docutilsParsersRstParser.tabWidth;
        const inputLines = statemachine.string2lines(
            inputstring, {
                tabWidth: tabWidth,
                convertWhitespace: true,
            },
        );

        //      console.log(`initial state is ${this.initialState}`);
        if(this.stateMachine.debug) {
            console.log('fo')
        }
        this.stateMachine.run({ inputLines, document,
            inputOffset: 0,
            inliner: this.inliner });
        this.finishParse();
    }

    public finishParse(): void {
    }

}

export { Parser as RSTParser };

export default Parser;
