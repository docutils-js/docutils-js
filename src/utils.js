
import * as nodes from './nodes';
import { InvalidArgumentsError, SystemMessage, UnimplementedError as Unimp } from './Exceptions'
//export { SystemMessge };
import { combining } from './utils/combining';

export function getTrimFootnoteRefSpace(settings) {
    /*    """
    Return whether or not to trim footnote space.

    If trim_footnote_reference_space is not None, return it.

    If trim_footnote_reference_space is None, return False unless the
    footnote reference style is 'superscript'.
    """*/
    return false;
}
/*
    if settings.trim_footnote_reference_space is None:
        return hasattr(settings, 'footnote_references') and \
               settings.footnote_references == 'superscript'
    else:
        return settings.trim_footnote_reference_space
*/

export function columnWidth(text) { //fixme
    return text.length;
}
export function isIterable(obj) {
    // checks for null and undefined
    /* instanbul ignore if */
  if (obj == null) {
    return false;
  }
  return typeof obj[Symbol.iterator] === 'function';
}
export const punctuation_chars = {
    openers: '"\\\'(<\\\\[{\\u0f3a\\u0f3c\\u169b\\u2045\\u207d\\u208d\\u2329\\u2768' +
           '\\u276a\\u276c\\u276e\\u2770\\u2772\\u2774\\u27c5\\u27e6\\u27e8\\u27ea' +
           '\\u27ec\\u27ee\\u2983\\u2985\\u2987\\u2989\\u298b\\u298d\\u298f\\u2991' +
           '\\u2993\\u2995\\u2997\\u29d8\\u29da\\u29fc\\u2e22\\u2e24\\u2e26\\u2e28' +
           '\\u3008\\u300a\\u300c\\u300e\\u3010\\u3014\\u3016\\u3018\\u301a\\u301d' +
           '\\u301d\\ufd3e\\ufe17\\ufe35\\ufe37\\ufe39\\ufe3b\\ufe3d\\ufe3f\\ufe41' +
           '\\ufe43\\ufe47\\ufe59\\ufe5b\\ufe5d\\uff08\\uff3b\\uff5b\\uff5f\\uff62' +
           '\\xab\\u2018\\u201c\\u2039\\u2e02\\u2e04\\u2e09\\u2e0c\\u2e1c\\u2e20' +
           '\\u201a\\u201e\\xbb\\u2019\\u201d\\u203a\\u2e03\\u2e05\\u2e0a\\u2e0d' +
        '\\u2e1d\\u2e21\\u201b\\u201f',
    closers: '"\\\')>\\\\]}\\u0f3b\\u0f3d\\u169c\\u2046\\u207e\\u208e\\u232a\\u2769' +
               '\\u276b\\u276d\\u276f\\u2771\\u2773\\u2775\\u27c6\\u27e7\\u27e9\\u27eb' +
               '\\u27ed\\u27ef\\u2984\\u2986\\u2988\\u298a\\u298c\\u298e\\u2990\\u2992' +
               '\\u2994\\u2996\\u2998\\u29d9\\u29db\\u29fd\\u2e23\\u2e25\\u2e27\\u2e29' +
               '\\u3009\\u300b\\u300d\\u300f\\u3011\\u3015\\u3017\\u3019\\u301b\\u301e' +
               '\\u301f\\ufd3f\\ufe18\\ufe36\\ufe38\\ufe3a\\ufe3c\\ufe3e\\ufe40\\ufe42' +
               '\\ufe44\\ufe48\\ufe5a\\ufe5c\\ufe5e\\uff09\\uff3d\\uff5d\\uff60\\uff63' +
               '\\xbb\\u2019\\u201d\\u203a\\u2e03\\u2e05\\u2e0a\\u2e0d\\u2e1d\\u2e21' +
               '\\u201b\\u201f\\xab\\u2018\\u201c\\u2039\\u2e02\\u2e04\\u2e09\\u2e0c' +
        '\\u2e1c\\u2e20\\u201a\\u201e'
}


export class Reporter {
    constructor(source, reportLevel, haltLevel, stream, debug, encoding,
		errorHandler='backslashreplace') {
	if(haltLevel === undefined) {
	    haltLevel = 4;
	}
	this.DEBUG_LEVEL = 0
	this.INFO_LEVEL = 1
	this.WARNING_LEVEL = 2
	this.ERROR_LEVEL = 3
	this.SEVERE_LEVEL = 4
	this.source = source
	this.errorHandler = errorHandler
	this.debugFlag = debug
	this.reportLevel = reportLevel
	this.haltLevel = haltLevel
	// fixme
	this.stream = stream
	this.encoding = encoding; //fixme
	this.observers = []
	this.maxLevel = -1
    }

    setConditions() {
	throw new Unimp();
    }

    /* need better system for arguments!! */
    systemMessage(level, message, children, kwargs) {
	if(children == undefined) {
	    children= []
	}
	if(!isIterable(children)) {
	    //throw new Error(`Children is not iterable ${children}`);
	    kwargs = children
	    children = []
	}
	if(kwargs === undefined) {
	    kwargs = {}
	}

	if(message instanceof Error) {
	    message = message.message;
	}

	const attributes = { ...kwargs };
	if('base_node' in kwargs) {
	    const [ source, line ] = getSourceLine(kwargs.base_node);
	    delete attributes.base_node;
	    if(source && !attributes.source) {
		attributes.source = source;
	    }
	    if(line && !attributes.line) {
		attributes.line = line;
	    }
	}
	if(!('source' in attributes)) {
	    //fixme
	}
	const msg = new nodes.system_message(message, children, attributes)
	if(this.stream) {
	    this.stream.write(msg.astext() + '\n');
	}
	if(this.stream && (level >= this.reportLevel ||
			   (this.debugFlag && level == this.DEBUG_LEVEL)
			   || level >= this.haltLevel)) {
	    this.stream.write(msg.astext() + '\n');
	}
	if(level >= this.haltLevel) {
	    throw new SystemMessage(msg, level);
	}
	if(level > this.debugLevel || this.debugFlag) {
	    this.notifyObservers(msg)
	}
	this.maxLevel = Math.max(level, this.maxLevel)
	return msg
    }

    notifyObservers(message) {
	this.observers.forEach(o => o(message));
    }

    attachObserver(observer) {
	this.observers.push(observer);
    }

    debug(...args) {
	if(this.debugFlag) {
	    return this.systemMessage(this.debugLevel, ...args);
	}
    }
    info(...args) {
	return this.systemMessage(this.INFO_LEVEL, ...args);
    }
    warning(...args) {
	return this.systemMessage(this.WARNING_LEVEL, ...args);
    }
    error(...args) {
	return this.systemMessage(this.ERROR_LEVEL, ...args);
    }
    severe(...args) {
	return this.systemMessage(this.SEVERE_LEVEL, ...args);
    }
}

function _getCallerFile() {
    var originalFunc = Error.prepareStackTrace;

    	let callerfile;
		let callerlineno;
    try {
        var err = new Error();

        Error.prepareStackTrace = function (err, stack) { return stack; };

	const x = err.stack.shift();
 	const currentfile = x.getFileName();
	const currentlineno = x.getLineNumber();
//	process.stderr.write(`${currentfile} ${currentlineno}\n`);

        while (err.stack.length) {
	const x2 = err.stack.shift();
        callerfile = x2.getFileName();
	callerlineno = x2.getLineNumber();

            if(currentfile !== callerfile) break;
            }
    } catch (e) {
        console.log(e);
    }

    Error.prepareStackTrace = originalFunc;

    return [callerfile, callerlineno];
}

export function newReporter({sourcePath}, settings) {
    return new Reporter(sourcePath, settings.reportLevel,
			settings.haltLevel,
			settings.warningStream, settings.debug,
			settings.error_encoding,
			settings.error_encoding_error_handler)
}

export function escape2null(text) {
    //"""Return a string with escape-backslashes converted to nulls."""
    const parts = []
    let start = 0;
    while(true) {
        const found = text.indexOf('\\', start);
        if(found === -1) {
            parts.push(text.substring(start));
            return parts.join('');
	}
        parts.push(text.substring(start, found));
        parts.push('\x00' + text.substring(found+1, found+2));
        start = found + 2               // skip character after escape
    }
}

export function unescape(text, restoreBackslashes=false, respectWhitespace=false) {
    /*
    Return a string with nulls removed or restored to backslashes.
    Backslash-escaped spaces are also removed.
    */
    // `respect_whitespace` is ignored (since introduction 2016-12-16)if(
	if(typeof text === 'undefined') {
		throw new Error();

	}
    if(restoreBackslashes) {
        return text.replace('\x00', '\\');
    } else {
	return ['\x00 ', '\x00\n', '\x00'].reduce((a, v) => { return a.split(v).join('') },text||'')
    }
}


export function newDocument({sourcePath}, settings) {
    const reporter = newReporter({ sourcePath }, settings );
    const document = new nodes.document( settings, reporter, '', [], { source: sourcePath });
    document.noteSource(sourcePath, -1);
    return document;
}

export function splitEscapedWhitespace(text) {
    /*    """
    Split `text` on escaped whitespace (null+space or null+newline).
    Return a list of strings.
    """*/
    const strings = text.split('\x00 ')
    const s = []
    for (const string of strings) {
	s.push(...string.split('\x00\n'));
    }
    return s;
}

export function columnIndicies(text) {
    const stringIndicies = new Array(text.length);
    for(let  i = 0; i < text.length; i++) {
	stringIndicies[i] = i;
    }
    findCombiningChars(text).forEach((index) => {
        stringIndicies[index] = undefined;
    });

    return stringIndicies.filter(i => typeof i !== 'undefined');
}

/*    """Indices of Unicode string `text` when skipping combining characters.

    >>> from docutils.utils import column_indices
    >>> column_indices(u'A t ab le ')
    [0, 1, 2, 4, 5, 7, 8]

    """*/
    // TODO: account for asian wide chars here instead of using dummy
    // replacements in the tableparser?
/*    string_indices = range(len(text))
    findCombiningChars(text).forEach((index) => {
        string_indices[index] = undefined;
    });
    return [i for i in string_indices if i is not None]
}
*/
export function findCombiningChars(text) {
/*"""Return indices of all combining chars in  Unicode string `text`.

    >>> from docutils.utils import find_combining_chars
    >>> find_combining_chars(u'A t ab le ')
    [3, 6, 9]

    """*/
    return text.split('').map((c, i) => {
	const r = combining[text.codePointAt(i)]
	return [r, i];
    }).filter(([r, i]) => r).map(([r, i]) => i);
}


export default {
    newDocument,
}

