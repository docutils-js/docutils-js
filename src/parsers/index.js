import Parser from '../Parser';
import * as rst from './restructuredtext';

function getParserClass(parserName) {
    if (parserName === 'restructuredtext') {
        return rst;
    }
    throw new Error('');
//    return require(`./${parserName}.js`).default;
}

export default {
    getParserClass,
    Parser,
};
