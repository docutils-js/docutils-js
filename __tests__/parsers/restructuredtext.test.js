import { Parser } from '../../src/parsers/restructuredtext';
import { newDocument } from '../../src/utils';

test.skip('1', () => {
    const p = new Parser({});
    const document = newDocument({ sourcePath: '' }, {});
    p.parse('* a bullet point', document);
    expect(document.toString()).toMatchSnapshot();
});
