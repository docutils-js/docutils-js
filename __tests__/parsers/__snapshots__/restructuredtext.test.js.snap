// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 1`] = `"<document source=\\"\\"><bullet_list bullet=\\"*\\"><list_item><paragraph>a bullet point</paragraph></list_item></bullet_list></document>"`;

exports[`readme rst 1`] = `
"<document source=\\"\\"><section ids=\\"docutils-js\\" names=\\"docutils-js\\"><title>Docutils-JS</title><field_list><field><field_name>Author</field_name><field_body><paragraph>Kay McCormick</paragraph></field_body></field><field><field_name>Contact</field_name><field_body><paragraph>kaym2038@gmail.com</paragraph></field_body></field><field><field_name>Date</field_name><field_body><paragraph>Thu 18 Apr 2019 06:40:03 PM PDT</paragraph></field_body></field><field><field_name>Copyright</field_name><field_body><paragraph>MIT License</paragraph></field_body></field></field_list><target ids=\\"docutils-js github repository\\" names=\\"docutils-js github repository\\" refuri=\\"http://github.com/kaymccormick/docutils-js.git\\"/><target ids=\\"docutils\\" names=\\"docutils\\" refuri=\\"http://docutils.sourceforge.net/\\"/><target ids=\\"docutils distribution\\" names=\\"docutils distribution\\" refuri=\\"http://docutils.sourceforge.net/#download\\"/><system_message line=\\"11\\"><paragraph>Using English fallback for directive contents</paragraph></system_message><comment>unimplemented directive contents</comment></section><section ids=\\"introduction\\" names=\\"introduction\\"><title>Introduction</title><section ids=\\"about\\\\ the\\\\ docutils\\" names=\\"about\\\\ the\\\\ docutils\\"><title>About the Docutils</title><paragraph>The <reference name=\\"Docutils\\" refname=\\"docutils\\">Docutils</reference> is a Python library for producing documentation based on
a text format called RST or reStructuredText. RST is used for
documenting many prominent software projects. While primarily in the
Python ecosystem, the linux kernel also uses RST for its
documentation.</paragraph></section><section ids=\\"about\\\\ this\\\\ software\\" names=\\"about\\\\ this\\\\ software\\"><title>About this software</title><paragraph>Welcome to the port of the Docutils to JavaScript. This package is
designed to enable reStructuredText (RST) processing (reading,
transforming, writing) in a JavaScript environment. It is intended to
work in a browser and in NodeJS.</paragraph></section></section><section ids=\\"setup\\\\ and\\\\ configuration\\" names=\\"setup\\\\ and\\\\ configuration\\"><title>Setup and configuration</title><paragraph>1. Clone the repository with <literal>git</literal>:</paragraph><block_quote><literal_block>$ git clone http://github.com/kaymccormick/docutils-js.git docutils-js</literal_block></block_quote><paragraph>2. Enter the source directory:</paragraph><block_quote><literal_block>$ cd docutils-js</literal_block></block_quote><paragraph>3. Install the dependencies using <literal>yarn</literal>:</paragraph><block_quote><literal_block>$ yarn install</literal_block></block_quote><paragraph>4. Run <literal>grunt</literal> to transpile the source code:</paragraph><block_quote><literal_block>$ yarn grunt</literal_block></block_quote><paragraph>5. The output of this command should be in the <literal>lib</literal> subdirectory. Then, optionally run the tests:</paragraph><block_quote><literal_block>$ yarn jest</literal_block></block_quote></section><section ids=\\"using\\\\ the\\\\ api\\" names=\\"using\\\\ the\\\\ api\\"><title>Using the API</title><paragraph>Take a look at tools/rst2pojo.js. The API is in flux, because a
straight port of the docutils interfaces is not appropriate for a JS
API. JS does not have native IO - this is provided via the host in
some way. Node provides its own <literal>fs</literal> module, and web browsers of
course have many other ways of getting <literal>RST</literal> input, from
<literal>XMLHttpRequest</literal>/<literal>fetch</literal> to extracting text from the current
document or a form input (e.g. <literal>textarea</literal>).</paragraph><paragraph>Further, moving the IO responsibilities up the stack ensures that
deferred/asynchronous execution is handled outside of the docutils-<reference name=\\"js\\" refname=\\"js\\">js</reference>
module itself, improving the developer experience.</paragraph><paragraph>Example: For now, the simplest method to use the software appears to be:</paragraph><literal_block>const parse = require(&apos;../lib/index&apos;).parse;
const PojoWriter = require(&apos;../lib/writers/pojo.js&apos;).default;
const docSource = &apos;However you get your document sources&apos;
const document = parse(docSource);</literal_block><paragraph>The variable <literal>document</literal> is now ready for further processing.</paragraph></section><section ids=\\"addendum\\" names=\\"addendum\\"><title>Addendum</title><paragraph>Remaining to be implemented:</paragraph><block_quote><bullet_list bullet=\\"*\\"><list_item><paragraph>Transformers (output is not currently processed as docutils would)</paragraph></list_item><list_item><paragraph>Directives (there are a couple of no-op directives in place, such as image
and contents.</paragraph></list_item></bullet_list></block_quote><paragraph>There are currently quite a number of bugs.</paragraph><block_quote><bullet_list bullet=\\"*\\"><list_item><paragraph>Problems with sections</paragraph></list_item><list_item><paragraph>Problems with tables</paragraph></list_item><list_item><paragraph>Lack of transformers and directives</paragraph></list_item><list_item><paragraph>Running rst2xml.js on sample documents produces documents that
&apos;work&apos;, sort-of, but which are significantly different from
docutils-python output.</paragraph></list_item></bullet_list></block_quote><paragraph>However, the project should be usable for people who want to
experiment, play around, or contribute through bug fixes or feature
enhancements.</paragraph></section></document>"
`;
